source/common/http/header_map_impl.cc

```diff
@@ -21,190 +21,103 @@ constexpr size_t MinDynamicCapacity{32};
+#define BUFFER_IN_VEC_GET absl::get<absl::InlinedVector<char, 128>>(buffer_)
+
 void HeaderString::append(const char* data, uint32_t size) {
-  switch (type_) {
-  case Type::Reference: {
-    // Rather than be too clever and optimize this uncommon case, we dynamically
-    // allocate and copy.
-    type_ = Type::Dynamic;
-    const uint64_t new_capacity = newCapacity(string_length_, size);
-    if (new_capacity > MinDynamicCapacity) {
-      validateCapacity(new_capacity);
-      dynamic_capacity_ = new_capacity;
-    } else {
-      dynamic_capacity_ = MinDynamicCapacity;
-    }
-    char* buf = static_cast<char*>(malloc(dynamic_capacity_));
-    RELEASE_ASSERT(buf != nullptr, "");
-    memcpy(buf, buffer_.ref_, string_length_);
-    buffer_.dynamic_ = buf;
-    break;
+  uint64_t new_capacity = static_cast<uint64_t>(size) + string_length_;
+  if (new_capacity < MinDynamicCapacity) {
+    new_capacity = MinDynamicCapacity;
   }
+  validateCapacity(new_capacity);
+  ASSERT(validHeaderString(absl::string_view(data, size)));
 
-  case Type::Inline: {
-    const uint64_t new_capacity = static_cast<uint64_t>(size) + string_length_;
-    if (new_capacity <= sizeof(inline_buffer_)) {
-      // Already inline and the new value fits in inline storage.
-      break;
-    }
-
-    FALLTHRU;
+  if (type() == Type::Reference) {
+    // Rather than be too clever and optimize this uncommon case, we switch to
+    // Inline mode and copy.
+    buffer_ = absl::InlinedVector<char, 128>(BUFFER_STR_VIEW_GET.data(),
+                                             BUFFER_STR_VIEW_GET.data() + string_length_);
   }
 
-  case Type::Dynamic: {
-    // We can get here either because we didn't fit in inline or we are already dynamic.
-    if (type_ == Type::Inline) {
-      const uint64_t new_capacity = newCapacity(string_length_, size);
-      validateCapacity(new_capacity);
-      buffer_.dynamic_ = static_cast<char*>(malloc(new_capacity));
-      RELEASE_ASSERT(buffer_.dynamic_ != nullptr, "");
-      memcpy(buffer_.dynamic_, inline_buffer_, string_length_);
-      dynamic_capacity_ = new_capacity;
-      type_ = Type::Dynamic;
-    } else {
-      if (size + string_length_ > dynamic_capacity_) {
-        const uint64_t new_capacity = newCapacity(string_length_, size);
-        validateCapacity(new_capacity);
-
-        // Need to reallocate.
-        dynamic_capacity_ = new_capacity;
-        buffer_.dynamic_ = static_cast<char*>(realloc(buffer_.dynamic_, dynamic_capacity_));
-        RELEASE_ASSERT(buffer_.dynamic_ != nullptr, "");
-      }
-    }
-  }
-  }
-  ASSERT(validHeaderString(absl::string_view(data, size)));
-  memcpy(buffer_.dynamic_ + string_length_, data, size);
+  BUFFER_IN_VEC_GET.reserve(new_capacity);
+  BUFFER_IN_VEC_GET.insert(BUFFER_IN_VEC_GET.begin() + string_length_, data, data + size);
   string_length_ += size;
 }
 
-void HeaderString::clear() {
-  switch (type_) {
+char* HeaderString::buffer() {
```

Comment:
I'm assuming this is for the safety of not modifying a const reference string, correct? We might just want to get rid of this non-const accessor entirely. 

I think the only call-site of this method is in the H/1 codec where we access the buffer and lower-case it. Can we just append lower-cased data to the HeaderString instead of lower-casing the appended data AND GET RID OF THIS METHOD FOREVER?!

Reply:
IIUC it is also being used by HeaderMapImpl::addViaMove

https://github.com/envoyproxy/envoy/blob/e176b307bd2ab475574fc7c4b0f92e8575129573/source/common/http/header_map_impl.cc#L394-L395

I think the next step is to refactor header_map, so I'll remember this comment, and see if we can remove the clear method.

Reply:
I meant the `buffer()` method!

Reply:
+1 please kill it if possible.

Reply:
I've removed the `buffer()` method.
When thinking how to add only lowercase strings, I was faced with 2 options, as the given `data` is const ( https://github.com/envoyproxy/envoy/blob/e176b307bd2ab475574fc7c4b0f92e8575129573/source/common/http/http1/codec_impl.cc#L473 )

1) Allocate a new buffer, copy `data`, apply the lowercase function, and append the outcome to the HeaderString (=> will copy the data twice).
2) Append the original data (that may require lowercasing), and apply a "transform" function on the HeaderString (=> will copy the data once, and then lowercase in-place).
I've decided to use option 2, and introduce a new method called `inlineTransform`, that calls std::transform on the data. It should use less memory.

Let me know if you think I should use option 1, or if you have a better idea.

Reply:
Ugh. Option (1) is pretty bad. It bothers me that there are at least 3 ways we `tolower` strings, including (1) `LowerCaseString` with ctype.h `tolower`, (2) the fast lookup `ToLowerTable` like here, and (3) `StringUtil::toLower` using `absl::ascii_tolower` (and also the `absl::AsciiStrToLower` string version)

Curious, is (3) fast? It seems to something similar to our lower table. CAN WE GET RID OF THE LOWER TABLE TOO AND JUST USE `absl::ascii_tolower` AS THE UNARY_OP?

Reply:
Welcome to leave as TODO since (2) gives the option of switching out the unary op anyway

Reply:
Good point. I'll change it to use `absl::ascii_tolower`, and get rid of the lower table.